tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/ParameterNormalizer.scala

```diff
@@ -0,0 +1,116 @@
+      else normalizedLetIn
+    val newDecl = decl.copy( body = newBody )
+    // Copy doesn't preserve .isRecursive!
+    newDecl.isRecursive = decl.isRecursive
+    newDecl
+  }
+
+  private def normalizeInternalLetIn : TlaExTransformation = tracker.track {
+    case ex@LetInEx( body, defs@_* ) =>
+      val newDefs = defs map { d => mkParamNormalForm( d ) }
+      val newBody = normalizeInternalLetIn( body )
+      if ( defs == newDefs && body == newBody ) ex
+      else LetInEx( newBody, newDefs : _* )
+
+    case ex@OperEx( op, args@_* ) =>
+      val newArgs = args map normalizeInternalLetIn
+      if ( args == newArgs ) ex
+      else OperEx( op, newArgs : _* )
+
+    case ex => ex
+
+  }
+
+  private def paramNormal( paramNames : List[FormalParam] ) : TlaExTransformation = tracker.track { ex =>
+    paramNames.foldLeft( ex ) {
+      case (partialEx, fParam) =>
+        // For each formal parameter we invent a fresh operator name
+        val paramOperName = nameGenerator.newName()
+        // We split now, based on whether fParam is a simple or operator parameter
+        fParam match {
+          case SimpleFormalParam( name ) =>
+            // We replace all instances of `fParam` with `paramOperName`
+            // however, since paramOperName is an operator, we have to replace with application
+            val tr = ReplaceFixed(
+              NameEx( fParam.name ),
+              OperEx( TlaOper.apply, NameEx( paramOperName ) ),
+              tracker
+            )
+            val replaced = tr( partialEx )
+            // if fParam is simple, the introduced operator is nullary
+            val letInDef = TlaOperDecl( paramOperName, List.empty, NameEx( name ) )
+            LetInEx( replaced, letInDef )
+
+          case OperFormalParam( name, arity ) =>
+            // We again replace all instances of `fParam` with `paramOperName`
+            // As both are operators, we don't need to introduce application
+            val tr = ReplaceFixed(
+              NameEx( fParam.name ),
+              NameEx( paramOperName ),
+              tracker
+            )
+            val replaced = tr( partialEx )
+
+            // This time, the introduced operator is not nullary, so we need to invent parameters
+            val inventedParams = List.fill( arity ) {
+              nameGenerator.newName()
+            }
+            // The body is just the operator applied to all the parameters
+            val newBody = OperEx(
+              TlaOper.apply, name +: inventedParams map NameEx : _*
+            )
+            val letInDef = TlaOperDecl( paramOperName, inventedParams map SimpleFormalParam, newBody )
+            LetInEx( replaced, letInDef )
+        }
+    }
+  }
+
+  def normalize : TlaModuleTransformation = { m =>
+    // Iterates over all declarations and replaces those for which decisionFn returns true
+    // (by default, for all recursive)
```

Comment:
Should this maybe be a document comment, since its quite informative re: this public method?